src/memory.jl

```diff
@@ -1,203 +1,318 @@
 # This file is a part of JuliaFEM.
 # License is MIT: see https://github.com/JuliaFEM/Materials.jl/blob/master/LICENSE
 
-@with_kw mutable struct MemoryDriverState <: AbstractMaterialState
-    time :: Float64 = zero(Float64)
-    strain :: SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
+module MemoryModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, lame, debang
+import ..integrate_material!  # for method extension
+
+# parametrically polymorphic for any type representing ℝ
+export GenericMemory, GenericMemoryDriverState, GenericMemoryParameterState, GenericMemoryVariableState
+
+# specialization for Float64
+export Memory, MemoryDriverState, MemoryParameterState, MemoryVariableState
+
+@with_kw mutable struct GenericMemoryDriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
+end
+
+# TODO: complete this docstring
+"""Parameter state for Memory material.
+
+`E`: Young's modulus
+`nu`: Poisson's ratio
+`R0`: initial yield strength
+`Kn`: plasticity multiplier divisor (drag stress)
+`nn`: plasticity multiplier exponent
+`C1`, `D1`: parameters governing behavior of backstress X1
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q0`: ???
+`QM`: ???
+`mu`: ???
+`b`: ???
+`eta`: ???
+`m`: ???
+`pt`: memory evanescence threshold for cumulative equivalent plastic strain
+"""
+@with_kw struct GenericMemoryParameterState{T <: Real} <: AbstractMaterialState
+    E::T = 0.0
+    nu::T = 0.0
+    R0::T = 0.0
+    Kn::T = 0.0
+    nn::T = 0.0
+    C1::T = 0.0
+    D1::T = 0.0
+    C2::T = 0.0
+    D2::T = 0.0
+    Q0::T = 0.0
+    QM::T = 0.0
+    mu::T = 0.0
+    b::T = 0.0
+    eta::T = 0.0
+    m::T = 0.0
+    pt::T = 0.0
```

Comment:
Threshold of equivalent plastic strain, after which the memory evanescence starts.